### PR TITLE
[SPARK-31959][SQL][test-java11] Fix Gregorian-Julian micros rebasing while switching standard time zone offset

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Fix the bug in microseconds rebasing during transitions from one standard time zone offset to another one. In the PR, I propose to change the implementation of `rebaseGregorianToJulianMicros` which performs rebasing via local timestamps. In the case of overlapping:
1. Check that the original instant belongs to earlier or later instant of overlapped local timestamp.
2. If it is an earlier instant, take zone and DST offsets from the previous day otherwise
3. Set time zone offsets to Julian timestamp from the next day.

Note: The fix assumes that transitions cannot happen more often than once per 2 days.

### Why are the changes needed?
Current implementation handles timestamps overlapping only during daylight saving time but overlapping can happen also during transition from one standard time zone to another one. For example in the case of `Asia/Hong_Kong`, the time zone switched from `Japan Standard Time` (UTC+9) to `Hong Kong Time` (UTC+8) on _Sunday, 18 November, 1945 01:59:59 AM_. The changes allow to handle the special case as well.

### Does this PR introduce _any_ user-facing change?
It might affect micros rebasing in before common era when not-optimised version of `rebaseGregorianToJulianMicros()` is used directly.

### How was this patch tested?
1. By existing tests in `DateTimeUtilsSuite`, `RebaseDateTimeSuite`, `DateFunctionsSuite`, `DateExpressionsSuite` and `TimestampFormatterSuite`.
2. Added new test to `RebaseDateTimeSuite`
3. Regenerated `gregorian-julian-rebase-micros.json` with the step of 30 minutes, and got the same JSON file. The JSON file isn't affected because previously it was generated with the step of 1 week. And the spike in diffs/switch points during 1 hour of timestamp overlapping wasn't detected.

### CODE REVIEW

**Comment by MaxGekk** [comment]
The build https://github.com/apache/spark/pull/28787#issuecomment-642292477 failed on the assert for JDK calls:
```scala
      val ldt = LocalDateTime.of(1945, 11, 18, 1, 30, 0)
      val earlierMicros = instantToMicros(ldt.atZone(hkZid).withEarlierOffsetAtOverlap().toInstant)
      val laterMicros = instantToMicros(ldt.atZone(hkZid).withLaterOffsetAtOverlap().toInstant)
      assert(earlierMicros + MICROS_PER_HOUR === laterMicros)
```
because jenkins uses "old" JDK:
```
JENKINS_MASTER_HOSTNAME=amp-jenkins-master
JAVA_HOME=/usr/java/jdk1.8.0_191
```
which has an outdated time zone database, see https://bugs.openjdk.java.net/browse/JDK-8228469
```
     Hong Kong's 1941-06-15 spring-forward transition was at 03:00, not
     03:30. Its 1945 transition from JST to HKT was on 11-18 at 02:00,
     not 09-15 at 00:00. In 1946 its spring-forward transition was on
     04-21 at 00:00, not the previous day at 03:30. From 1946 through
     1952 its fall-back transitions occurred at 04:30, not at 03:30.
     In 1947 its fall-back transition was on 11-30, not 12-30.
     (Thanks to P Chan.)
```
@cloud-fan @HyukjinKwon @dongjoon-hyun @srowen Can we upgrade JDK on jenkins machines?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala [hunk 1]
```diff
@@ -409,4 +409,31 @@ class RebaseDateTimeSuite extends SparkFunSuite with Matchers with SQLHelper {
...
+      val laterMicros = instantToMicros(ldt.atZone(hkZid).withLaterOffsetAtOverlap().toInstant)
+      assert(earlierMicros + MICROS_PER_HOUR === laterMicros)
+      val rebasedEarlierMicros = rebaseGregorianToJulianMicros(hkZid, earlierMicros)
+      val rebasedLaterMicros = rebaseGregorianToJulianMicros(hkZid, laterMicros)
+      def toTsStr(micros: Long): String = toJavaTimestamp(micros).toString
+      val expected = "1945-11-18 01:30:00.0"
+      assert(toTsStr(rebasedEarlierMicros) === expected)
+      assert(toTsStr(rebasedLaterMicros) === expected)
+      assert(rebasedEarlierMicros + MICROS_PER_HOUR === rebasedLaterMicros)
+      // Check optimized rebasing
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala:431]: what do you mean by "optimized rebase"?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala [hunk 2]
```diff
@@ -326,20 +326,34 @@ object RebaseDateTime {
...
-      .set(Calendar.DST_OFFSET, zoneId.getRules.getDaylightSavings(instant).toMillis.toInt)
       .build()
+    // A local timestamp can have 2 instants in the cases of switching from:
+    //  1. Summer to winter time.
+    //  2. One standard time zone to another one. For example, Asia/Hong_Kong switched from JST
+    //     to HKT on 18 November, 1945 01:59:59 AM.
+    // Below we check that the original `instant` is earlier or later instant. If it is an earlier
+    // instant, we take the standard and DST offsets of the previous day otherwise of the next one.
+    val trans = zoneId.getRules.getTransition(ldt)
+    if (trans != null && trans.isOverlap) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:347]: when will we go into this expensive branch?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:347]: I wrote the comment above

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala:431]: via pre-calculated offsets and switch points.

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala:431]: so this relies on the JVM system timezone is not `Asia/Hong_Kong`?

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala:431]: Yes, it relies but I set the default JVM time zone to `Asia/Hong_Kong` in the test.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by cloud-fan** [comment]
More context:

To do datetime rebase more efficiently, we put the diffs between two calendars in JSON files, to avoid calculating the diffs every time.

We run the rebase function for all the datetime values from year 0001 to year 2037, using one week as the step. This is not accurate and we miss diffs during some short period.

The good news is: although the JSON files do not match the rebase function, they are correct. The rebase function has a bug. We have a random test to make sure the rebase function matches the generated JSON files, and that's why we found this bug.

That said, this is not a real bug as it doesn't affect end users, but it makes the test flaky.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by dongjoon-hyun** [comment]
Hi, All. As I commented on the backporting PR, this causes a weird consistent failure on `research-jenkins-worker-09`. Could you take a look at that?
- https://github.com/apache/spark/pull/28809#discussion_r439869464
- https://github.com/apache/spark/pull/28809#pullrequestreview-430253537

**Pull request closed without merging**